### PR TITLE
Reinit shape functions at physical points within neighbor element

### LINKED ISSUE #6607: reinit shape functions at physical points within neighbor element
### Description of the enhancement or error report
This will add and fix some functions to allow users to reinit shape functions at physical points within neighbor element.  

### Rationale for the enhancement or information for reproducing the error
I tried to reinit shape functions at physical points within neighbor element for XFEM constraint problem. From doing this, I observed some inconsistency of several functions in the framework. 

1. In ```Assembly.C```, if you want to use ```reinitNeighborAtPhysical```, you have to pass in the neighbor side even though the physical points are not necessarily on the side.

2. In ```MooseVariable.C```, the codes of ```computeNeighborValuesFace``` and ```computeNeighborValues``` are identical because only the 'shape functions on the face' are calculated and stored.

3. ```MooseVariable.C``` only provides ```phiFaceNeighbor()...``` functions even though they can return the shape functions at the points within the element. It is a bit confusing.     

I understand that the existing functions are designed for more general purpose which I probably did not realize. Here I just tried to modify some functions based on my best knowledge. My code change very likely violates some of your design purpose and just let me know.  

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
Have impact to all neighbor reinit related codes, such as DG, constraint, Interface. I made my code change be able to pass all the regression test.

### PR DESCRIPTION
**Pull request by jiangwen84**
This will add and fix some functions to allow users to reinit shape functions at physical points within neighbor element.  These functions are used in `ElemElemConstraint` and `NonlinearSystem`, see recent merged #6756 .

The existing tests under `\module\xfem\tests\single_var_constraint_2d` and `single_var_constraint_3d` are used to test the changed codes. 

closes #6607

### CODE REVIEW

**Comment by permcody** [comment]
There's a lot here in Assembly and other low level classes. @andrsd you should probably review this.

**Comment by jiangwen84** [comment]
@andrsd and @jwpeterson helped me look at the codes before. Could one of you help me review this code again? Thanks!

File: framework/src/base/Assembly.C [hunk 1]
```diff
@@ -69,6 +69,7 @@ Assembly::Assembly(SystemBase & sys, CouplingMatrix * & cm, THREAD_ID tid) :
   // Build fe's for the helpers
   buildFE(FEType(FIRST, LAGRANGE));
   buildFaceFE(FEType(FIRST, LAGRANGE));
+  buildNeighborFE(FEType(FIRST,LAGRANGE));
```
Comment (andrsd) [inline framework/src/base/Assembly.C]: Space after `,`

File: framework/src/base/Assembly.C [hunk 2]
```diff
@@ -105,6 +106,9 @@ Assembly::~Assembly()
   for (unsigned int dim=1; dim<=_mesh_dimension; dim++)
     for (std::map<FEType, FEBase *>::iterator it = _fe_neighbor[dim].begin(); it != _fe_neighbor[dim].end(); ++it)
       delete it->second;
+  for (unsigned int dim=1; dim<=_mesh_dimension; dim++)
```
Comment (andrsd) [inline framework/src/base/Assembly.C]: Spaces around operators

File: framework/src/base/Assembly.C [hunk 3]
```diff
@@ -191,6 +195,24 @@ Assembly::buildFaceNeighborFE(FEType type)
...
 }
 
+void
+Assembly::buildFaceNeighborFE(FEType type)
+{
+  if (!_fe_shape_data_face_neighbor[type])
+    _fe_shape_data_face_neighbor[type] = new FEShapeData;
+
+  // Build an FE object for this type for each dimension up to the dimension of the current mesh
+  for (unsigned int dim=1; dim<=_mesh_dimension; dim++)
```
Comment (andrsd) [inline framework/src/base/Assembly.C]: Spaces around operators

File: framework/src/base/Assembly.C [hunk 4]
```diff
@@ -715,17 +789,34 @@ Assembly::reinitNeighborAtPhysical(const Elem * neighbor, unsigned int neighbor_
...
+  unsigned int neighbor_dim = neighbor->dim();
+  FEInterface::inverse_map(neighbor_dim, FEType(), neighbor, physical_points, reference_points);
 
+  reinitFENeighbor(neighbor, reference_points);
+  reinitNeighbor(neighbor, reference_points);
   // Save off the physical points
   _current_physical_points = physical_points;
 }
 
+
```
Comment (andrsd) [inline framework/src/base/Assembly.C]: Nope ;-)

File: framework/src/base/DisplacedProblem.C [hunk 5]
```diff
@@ -378,9 +378,25 @@ DisplacedProblem::reinitNeighborPhys(const Elem * neighbor, unsigned int neighbo
...
+  // Set the neighbor dof indices
+  _displaced_nl.prepareNeighbor(tid);
+  _displaced_aux.prepareNeighbor(tid);
+
+  prepareAssemblyNeighbor(tid);
+
+  // Compute values at the points
   _displaced_nl.reinitNeighbor(neighbor, tid);
-  _displaced_aux.reinitNeighbor(neighbor,tid);
+  _displaced_aux.reinitNeighbor(neighbor, tid);
```
Comment (andrsd) [inline framework/src/base/DisplacedProblem.C]: 👍

File: framework/src/base/FEProblem.C [hunk 6]
```diff
@@ -1150,14 +1150,37 @@ FEProblem::reinitNeighborPhys(const Elem * neighbor, unsigned int neighbor_side,
...
   _nl.reinitNeighbor(neighbor, tid);
   _aux.reinitNeighbor(neighbor, tid);
 
   // Do the same for the displaced problem
   if (_displaced_problem != NULL)
-    _displaced_problem->reinitNeighborPhys(_displaced_mesh->elem(neighbor->id()), neighbor_side, physical_points, tid);
+    _displaced_problem->reinitNeighborPhys(_displaced_mesh->elem(neighbor->id()), physical_points, tid);
 }
 
+
```
Comment (andrsd) [inline framework/src/base/FEProblem.C]: Nope...

**Comment by andrsd** [comment]
Design-wise, it looks good

**Comment by jiangwen84** [comment]
thanks @andrsd  :-)

**Comment by permcody** [comment]
👍

File: framework/src/base/Assembly.C [hunk 7]
```diff
@@ -174,13 +178,13 @@ Assembly::buildFaceFE(FEType type)
 }
 
 void
-Assembly::buildFaceNeighborFE(FEType type)
+Assembly::buildNeighborFE(FEType type)
 {
-  if (!_fe_shape_data_face_neighbor[type])
-    _fe_shape_data_face_neighbor[type] = new FEShapeData;
+  if (!_fe_shape_data_neighbor[type])
+    _fe_shape_data_neighbor[type] = new FEShapeData;
```
Comment (permcody) [inline framework/src/base/Assembly.C]: @jiangwen84 - It looks like you didn't clean up this structure. I think that's your only problem.

**Pull request merged**